src/v2/params/types.ts

```diff
@@ -69,31 +72,36 @@ export interface SelectOptions<T = unknown> {
   value: T;
 }
 
-export interface ParamSpec<T = unknown> {
+export type ParamSpec<T = unknown> = {
   name: string;
   default?: T;
   label?: string;
-  as?: string;
   description?: string;
   type: ParamValueType;
-  input?: ParamInput<T>;
-}
+} & HasAtMostOneInput<T>;
 
 export type ParamOptions<T = unknown> = Omit<ParamSpec<T>, 'name' | 'type'>;
 
-export class Param<T = unknown> {
+export abstract class Param<T extends string | number | boolean | string[]> {
   static type: ParamValueType = 'string';
 
   constructor(readonly name: string, readonly options: ParamOptions<T> = {}) {}
 
-  get rawValue(): string | undefined {
-    return process.env[this.name];
+  get value(): T {
+    throw new Error('Not implemented');
+  }
+
+  expr() {
+    return new ParamExpression<T>(this);
   }
 
-  get value(): any {
-    return this.rawValue || this.options.default || '';
+  cmp(cmp: '==' | '>' | '>=' | '<' | '<=', rhs: T) {
+    return new CompareExpression<T>(cmp, this.expr(), rhs);
   }
 
+  equals(rhs: T) {
+    return this.cmp('==', rhs);
+  }
   toString() {
```

Comment:
As mentioned before, this versoin of `toString` may not work because you lose easy access to the part of the expression that doesn't have the {{}} escape sequence. That's what `expr` was for. Every Expression subclass implements its own `expr` (operators do so recursively) and then `toString` and `toJSON` are simply \`{{ ${ this.expr() } }}\`

Reply:
ack (cf3-params-detailed)